### PR TITLE
added StarShaped domain solves #1280

### LINKED ISSUE #1280: Add star-shaped geometries in 2D
It would be interesting to add a general star-shaped geometry. For star-shaped domains, one can set the radius as:
```
r(\theta) = r_0 + \sum_{j=1}^J [a_j\sin(j \theta)  + b_j \cos(j \theta)]
```
The input parameters could be: `(r_0, J, a_j, b_j)`.

I'll try to cook up something. 

https://github.com/lululxvi/deepxde/discussions/1264

### PR DESCRIPTION
**Pull request by pescap**

### CODE REVIEW

File: deepxde/geometry/geometry_2d.py [hunk 1]
```diff
@@ -264,6 +265,97 @@ def is_valid(vertices):
         )
 
 
+class StarShaped(Geometry):
+    """Star-shaped 2d domain.
```
Comment (lululxvi) [inline deepxde/geometry/geometry_2d.py:269]: Need a clear definition of "Star-shaped"

Comment (pescap) [reply deepxde/geometry/geometry_2d.py:269]: Hi, I am using the parametrization in:
[Large deformation shape uncertainty quantification in acoustic scattering](https://link.springer.com/article/10.1007/s10444-018-9594-8) (refer to Fig. 1 and Eq. 2.1).


There might be a general name for these Fourier expansion-based parametric domains... FourierDomain?

File: examples/pinn_forward/Schrodinger.ipynb [hunk 2]
```diff
@@ -38,17 +38,17 @@
      "text": [
       "Collecting deepxde\n",
       "  Downloading DeepXDE-0.14.0-py3-none-any.whl (111 kB)\n",
-      "\u001b[?25l\r\n",
```
Comment (lululxvi) [inline examples/pinn_forward/Schrodinger.ipynb:41]: Why updating this file?

File: deepxde/geometry/geometry_2d.py [hunk 3]
```diff
@@ -264,6 +264,104 @@ def is_valid(vertices):
         )
 
 
+class StarShaped(Geometry):
+    """Star-shaped 2d domain. Computes a geometry whose boundary is parametrized in polar coordinates as:
```
Comment (lululxvi) [inline deepxde/geometry/geometry_2d.py:268]: Star-shaped 2d domain, i.e., a geometry whose boundary is parametrized in polar coordinates as:

File: deepxde/geometry/geometry_2d.py [hunk 4]
```diff
@@ -264,6 +264,104 @@ def is_valid(vertices):
...
+    
+    $$
+    r(theta) := r_0 + sum_{i = 1}^N [a_i cos( i theta) + b_i sin(i theta) ],  theta in [0,2 pi].
+    $$
+    
+    For more details, refer to Large deformation shape uncertainty quantification in acoustic scattering, 
+    Hiptmair et. al., Adv. Comp. Math. (2018).
```
Comment (lululxvi) [inline deepxde/geometry/geometry_2d.py:275]: Add the link here.

File: deepxde/geometry/geometry_2d.py [hunk 5]
```diff
@@ -264,6 +264,105 @@ def is_valid(vertices):
...
+class StarShaped(Geometry):
+    """Star-shaped 2d domain, i.e., a geometry whose boundary is parametrized in polar coordinates as:
+    
+    $$
+    r(theta) := r_0 + sum_{i = 1}^N [a_i cos( i theta) + b_i sin(i theta) ],  theta in [0,2 pi].
+    $$
+    
+    For more details, refer to Large deformation shape uncertainty quantification in acoustic scattering, 
+    Hiptmair et. al., Adv. Comp. Math. (2018). 
+    Link: https://link.springer.com/article/10.1007/s10444-018-9594-8
```
Comment (lululxvi) [inline deepxde/geometry/geometry_2d.py:276]: Add link to the text, e.g., 

https://github.com/lululxvi/deepxde/blob/d6e60d19cf98e127fef25435837eca791b302e32/deepxde/nn/tensorflow_compat_v1/deeponet.py#L15

File: deepxde/geometry/geometry_2d.py [hunk 6]
```diff
@@ -264,6 +264,105 @@ def is_valid(vertices):
...
+        self.max_radius = radius + np.sum(coeffs_cos) + np.sum(coeffs_sin)
+        super().__init__(
+            2,
+            (self.center - self.max_radius, self.center + self.max_radius),
+            2 * self.max_radius,
+        )
+
+        self._r2 = radius**2
+
+    def r_theta(self, theta):
```
Comment (lululxvi) [inline deepxde/geometry/geometry_2d.py:299]: `def _r_theta(self, theta):`

File: deepxde/geometry/geometry_2d.py [hunk 7]
```diff
@@ -264,6 +264,105 @@ def is_valid(vertices):
...
+        """Define the parametrization r(theta) at angles theta."""
+        result = self.radius * np.ones(theta.shape)
+        for i, (coeff_cos, coeff_sin) in enumerate(
+            zip(self.coeffs_cos, self.coeffs_sin), start=1
+        ):
+            result += coeff_cos * np.cos(i * theta) + coeff_sin * np.sin(i * theta)
+        return result
+
+    def dr_theta(self, theta):
```
Comment (lululxvi) [inline deepxde/geometry/geometry_2d.py:308]: `def _dr_theta(self, theta):`

File: deepxde/geometry/geometry_2d.py [hunk 8]
```diff
@@ -264,6 +264,106 @@ def is_valid(vertices):
...
+        coeffs_cos: i-th order coefficients for the i-th cos term (a_i).
+        coeffs_sin: i-th order coefficients for the i-th sin term (b_i).
+    """
+
+    def __init__(self, center, radius, coeffs_cos, coeffs_sin):
+        self.center = np.array(center, dtype=config.real(np))
+        self.radius = radius
+        self.coeffs_cos = coeffs_cos
+        self.coeffs_sin = coeffs_sin
+        self.max_radius = radius + np.sum(coeffs_cos) + np.sum(coeffs_sin)
```
Comment (lululxvi) [inline deepxde/geometry/geometry_2d.py:291]: `self.max_radius` --> `max_radius`

File: deepxde/geometry/geometry_2d.py [hunk 9]
```diff
@@ -264,6 +264,106 @@ def is_valid(vertices):
...
+        super().__init__(
+            2,
+            (self.center - self.max_radius, self.center + self.max_radius),
+            2 * self.max_radius,
+        )
+
+        self._r2 = radius**2
```
Comment (lululxvi) [inline deepxde/geometry/geometry_2d.py:298]: `self._r2` never used.

**Pull request merged**